### PR TITLE
Add database methods for optimistic sync

### PR DESCRIPTION
**Pull request by potuz**
This PR adds the logic to fallback to DB the check of an optimistic status of a block that is no longer in Forkchoice. 
cc @jmozah 

Notice that this suggests a replacement of #10351 just to make it easier to write the remainder PRs

### CODE REVIEW

File: beacon-chain/blockchain/chain_info.go [hunk 1]
```diff
@@ -323,13 +325,34 @@ func (s *Service) IsOptimistic(ctx context.Context) (bool, error) {
...
+	}
+	if err != protoarray.ErrUnknownNodeRoot && err != doublylinkedtree.ErrNilNode {
+		return false, err
+	}
+	blk, err := s.cfg.BeaconDB.Block(ctx, root)
+	if err != nil {
+		return false, err
+	}
+
+	validatedCheckpoint, err := s.cfg.BeaconDB.LastValidatedCheckpoint(ctx)
```
Comment (terencechain) [inline beacon-chain/blockchain/chain_info.go:346]: Why not just have `LastValidatedRoot` if we aren't using `validatedCheckpoint.Epoch` anyway? 😄

Comment (potuz) [reply beacon-chain/blockchain/chain_info.go:346]: Added a check for the epoch.

File: beacon-chain/blockchain/chain_info.go [hunk 2]
```diff
@@ -323,13 +325,43 @@ func (s *Service) IsOptimistic(ctx context.Context) (bool, error) {
...
 
 // IsOptimisticForRoot takes the root and slot as aguments instead of the current head
 // and returns true if it is optimistic.
 func (s *Service) IsOptimisticForRoot(ctx context.Context, root [32]byte) (bool, error) {
-	return s.cfg.ForkChoiceStore.IsOptimistic(ctx, root)
+	optimistic, err := s.cfg.ForkChoiceStore.IsOptimistic(ctx, root)
+	if err == nil {
+		return optimistic, nil
+	}
+	if err != protoarray.ErrUnknownNodeRoot && err != doublylinkedtree.ErrNilNode {
```
Comment (terencechain) [inline beacon-chain/blockchain/chain_info.go:338]: Are you sure `ErrNilNode` is required here? I'd prefer to only have `ErrUnknownNodeRoot`

File: beacon-chain/blockchain/chain_info.go [hunk 3]
```diff
@@ -323,13 +325,43 @@ func (s *Service) IsOptimistic(ctx context.Context) (bool, error) {
...
+		return false, err
+	}
+	blk, err := s.cfg.BeaconDB.Block(ctx, root)
+	if err != nil {
+		return false, err
+	}
```
[the remark this answers is not in the document]
Comment (potuz) [reply beacon-chain/blockchain/chain_info.go:344]: how do you get `slot` without getting the Block?

**Review by terencechain — DISMISSED** [review]

**Pull request merged**